File: nativelib/src/main/resources/posix/sys/resource.c
#include <sys/resource.h>

struct scalanative_rlimit {
    rlim_t rlim_cur;
    rlim_t rlim_max;
};

struct scalanative_rusage {
    struct timeval ru_utime;
    struct timeval ru_stime;
};

static void convert_from_scalanative_rlimit(struct scalanative_rlimit *in,
                                            struct rlimit *out) {
    out->rlim_cur = in->rlim_cur;
    out->rlim_max = in->rlim_max;
}

static void convert_to_scalanative_rlimit(struct rlimit *in,
                                          struct scalanative_rlimit *out) {
    out->rlim_cur = in->rlim_cur;
    out->rlim_max = in->rlim_max;
}

static void convert_to_scalanative_rusage(struct rusage *in,
                                          struct scalanative_rusage *out) {
    out->ru_utime = in->ru_utime;
    out->ru_stime = in->ru_stime;
}

int scalanative_getrlimit(int resource, struct scalanative_rlimit *snRlim) {
    struct rlimit rlim;

    int status = getrlimit(resource, &rlim);

    if (status >= 0) {
        convert_to_scalanative_rlimit(&rlim, snRlim);
    } // contents at snRlim undefined on error.

    return status;
}

int scalanative_getrusage(int who, struct scalanative_rusage *snUsage) {
    struct rusage usage;

    int status = getrusage(who, &usage);

    if (status >= 0) {
        convert_to_scalanative_rusage(&usage, snUsage);
    } // contents at snUsage undefined on error.

    return status;
}

int scalanative_setrlimit(int resource, struct scalanative_rlimit *snRlim) {

    struct rlimit rlim;

    convert_from_scalanative_rlimit(snRlim, &rlim);

    int status = setrlimit(resource, &rlim);

    return status;
}

int scalanative_PRIO_PROCESS() { return PRIO_PROCESS; };

int scalanative_PRIO_PGRP() { return PRIO_PGRP; };

int scalanative_PRIO_USER() { return PRIO_USER; };

rlim_t scalanative_RLIM_INFINITY() { return RLIM_INFINITY; };

rlim_t scalanative_RLIM_SAVED_CUR() { return RLIM_SAVED_CUR; };

rlim_t scalanative_RLIM_SAVED_MAX() { return RLIM_SAVED_MAX; };

int scalanative_RLIMIT_AS() { return RLIMIT_AS; };

int scalanative_RLIMIT_CORE() { return RLIMIT_CORE; };

int scalanative_RLIMIT_CPU() { return RLIMIT_CPU; };

int scalanative_RLIMIT_DATA() { return RLIMIT_DATA; };

int scalanative_RLIMIT_FSIZE() { return RLIMIT_FSIZE; };

int scalanative_RLIMIT_NOFILE() { return RLIMIT_NOFILE; };

int scalanative_RLIMIT_STACK() { return RLIMIT_STACK; };

int scalanative_RUSAGE_CHILDREN() { return RUSAGE_CHILDREN; };

int scalanative_RUSAGE_SELF() { return RUSAGE_SELF; };

File: posixlib/src/main/scala/scala/scalanative/posix/sys/resource.scala
package scala.scalanative
package posix
package sys

// Reference:
//   The Open Group Base Specifications Issue 7, 2018 edition
//   https://pubs.opengroup.org/onlinepubs/9699919799/basedefs\
//       /sys_resource.h.html
//
//   Method argument names come from Ubuntu 19.04 linux man pages.
//   Open Group seems to no longer suggest them.

import scalanative.unsafe.*

@extern
@define("__SCALANATIVE_POSIX_SYS_RESOURCE")
object resource {

  type id_t = sys.types.id_t

  type rlim_t = CUnsignedLongInt

  type timeval = sys.time.timeval

  type rlimit = CStruct2[
    rlim_t, // rlim_cur
    rlim_t // rlim_max
  ]

  type rusage = CStruct2[
    timeval, // ru_utime
    timeval // ru_stime
  ]

  def getpriority(which: CInt, who: id_t): CInt = extern

  @name("scalanative_getrlimit")
  def getrlimit(resource: CInt, rlim: Ptr[rlimit]): CInt = extern

  @name("scalanative_getrusage")
  def getrusage(who: CInt, usage: Ptr[rusage]): CInt = extern

  def setpriority(which: CInt, who: id_t, prio: CInt): CInt = extern

  @name("scalanative_setrlimit")
  def setrlimit(resource: CInt, rlim: Ptr[rlimit]): CInt = extern

  // Constants
  @name("scalanative_prio_process")
  def PRIO_PROCESS: CInt = extern

  @name("scalanative_prio_pgrp")
  def PRIO_PGRP: CInt = extern

  @name("scalanative_prio_user")
  def PRIO_USER: CInt = extern

  @name("scalanative_rlim_infinity")
  def RLIM_INFINITY: rlim_t = extern

  @name("scalanative_rlim_saved_cur")
  def RLIM_SAVED_CUR: rlim_t = extern

  @name("scalanative_rlim_saved_max")
  def RLIM_SAVED_MAX: rlim_t = extern

  @name("scalanative_rlimit_as")
  def RLIMIT_AS: CInt = extern

  @name("scalanative_rlimit_core")
  def RLIMIT_CORE: CInt = extern

  @name("scalanative_rlimit_cpu")
  def RLIMIT_CPU: CInt = extern

  @name("scalanative_rlimit_data")
  def RLIMIT_DATA: CInt = extern

  @name("scalanative_rlimit_fsize")
  def RLIMIT_FSIZE: CInt = extern

  @name("scalanative_rlimit_nofile")
  def RLIMIT_NOFILE: CInt = extern

  @name("scalanative_rlimit_stack")
  def RLIMIT_STACK: CInt = extern

  @name("scalanative_rusage_children")
  def RUSAGE_CHILDREN: CInt = extern

  @name("scalanative_rusage_self")
  def RUSAGE_SELF: CInt = extern
}

object resourceOps {

  import resource.{rlimit, rlim_t, rusage, timeval}

  implicit class rlimitOps(val ptr: Ptr[rlimit]) extends AnyVal {
    def rlim_cur: rlim_t = ptr._1
    def rlim_max: rlim_t = ptr._2
    def rlim_cur_=(v: rlim_t): Unit = ptr._1 = v
    def rlim_max_=(v: rlim_t): Unit = ptr._2 = v
  }

  implicit class rusageOps(val ptr: Ptr[rusage]) extends AnyVal {
    def ru_utime: timeval = ptr._1
    def ru_stime: timeval = ptr._2
    def ru_utime_=(v: timeval): Unit = ptr._1 = v
    def ru_stime_=(v: timeval): Unit = ptr._2 = v
  }
}
